tests/test_syntax/blocks/test_html_blocks.py

```diff
@@ -0,0 +1,783 @@
+                <!--
+
+                    *foo*
+
+                -->
+                """
+            ),
+            self.dedent(
+                """
+                <!--
+
+                    *foo*
+
+                -->
+                """
+            )
+        )
+
+    def test_raw_comment_nested(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <div>
+                <!-- *foo* -->
+                </div>
+                """
+            ),
+            self.dedent(
+                """
+                <div>
+                <!-- *foo* -->
+                </div>
+                """
+            )
+        )
+
+    def test_raw_processing_instruction_one_line(self):
+        self.assertMarkdownRenders(
+            "<?php echo '>';' ?>",
+            "<?php echo '>';' ?>"
+        )
+
+    # This is inline as it is not on a line by itself.
+    def test_raw_processing_instruction_one_line_followed_by_text(self):
+        self.assertMarkdownRenders(
+            "<?php echo '>';' ?>*bar*",
+            "<p><?php echo '>'; ' ?><em>bar</em></p>"
+        )
+
+    def test_raw_multiline_processing_instruction(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <?php
+                echo '>';'
+                ?>
+                """
+            ),
+            self.dedent(
+                """
+                <?php
+                echo '>';'
+                ?>
+                """
+            )
+        )
+
+    def test_raw_processing_instruction_with_blank_lines(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <?php
+
+                echo '>';'
+
+                ?>
+                """
+            ),
+            self.dedent(
+                """
+                <?php
+
+                echo '>';'
+
+                ?>
+                """
+            )
+        )
+
+    def test_raw_processing_instruction_indented(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <?php
+
+                    echo '>';'
+
+                ?>
+                """
+            ),
+            self.dedent(
+                """
+                <?php
+
+                    echo '>';'
+
+                ?>
+                """
+            )
+        )
+
+    def test_raw_declaration_one_line(self):
+        self.assertMarkdownRenders(
+            '<!DOCTYPE html>',
+            '<!DOCTYPE html>'
+        )
+
+    # TODO: Decide correct behavior. This matches current behavior and Commonmark.
+    # The reference implementation considers this inline not block level:
+    #
+    #   <p><!DOCTYPE html><em>bar</em></p>
+    #
+    # But most implementations do this instead:
+    #
+    #   <p>&lt;!DOCTYPE html&gt;<em>bar</em></p>
+    #
+    # Either makes sense, but the later seems more correct to me.
```

Comment:
I was going to say it is at the beginning, but that's not true as the opening `p` tag would be at the beginning. Good point. Thanks.

Reply:
I agree with @facelessuser here. We shouldn't be special casing different content within the tag. Anything that matches `<!.*>` should all get the same treatment (whether that's a doctype, cdata or anything else). The question is: what treatment? I think it is clear that anything between the angle brackets gets ignored by Markdown (so no escaping), but do we treat the tag as inline or block level? If it is inline, then it gets wraped in `<p>` tags, if block-level it doesn't.

But this issue, like others I raised, is specific to the behavior when the tag is followed by additional text on the same line. Based on other tests, I fully intend for `<!anything>` to be treated as blocklevel. But what about `<!anything>more text` where "more text" is on the same line. That is the only question I am asking here. Should that specific edge case be inline or block level?

I ask about this specific edge case because the existing behavior is odd. It treats the tag as blocklevel, but also skips Markdown parsing on the "more text". Interestingly, that is what the [Commonmark spec][2] requires. But we are not a Commonmark parser. So, should we follow the reference implementation and treat the tag as inline or something else?

Personally I would have expected "more text" to be broken out into a separate Markdown paragraph. But according to [Babelmark][1], almost noone does that (marked being the sole exception).

[1]: http://johnmacfarlane.net/babelmark2/?normalize=1&text=%3C!DOCTYPE+html%3E*bar*
[2]: https://spec.commonmark.org/0.28/#example-144

Reply:
I'm fine with any behavior. If you have no preference then maybe keep the existing one?